### DISCUSSION TITLE
Reduce build time for Android App in Release

### DISCUSSION THREAD

**Discussion by FlavioGoncalves-Cayas** [dotnet/maui#24689] [Q&A]
Hi everyone,
for one of our customers I am developing an app that will be whitelabeled into many visually different apps with slightly different configuration with the same codebase during build-time. 
Each app has it's own Pipeline in Azure DevOps for iOS and Android, where the build app step alone for Android takes about 10-15 minutes. I assume most of the time is spent during conversion from .dll to .so files. Now take that times >20 (apps) this means the build process for all apps takes several hours.

Is there any way to reduce that time since the codebase itself is mostly the same - only some configuration files, .csproj properties, resources etc. are switched out before building?

Would it be beneficial to put most of the shared code in it's own NuGet Package/DLL?
Or is there some way to build .so files in advance in a seperate pipeline and feed into the build for the multiple apps?

Any ideas?

Tagging @jonathanpeppers @jpobst for visibility.

**Comment by jonathanpeppers** [comment]
@FlavioGoncalves-Cayas can you "diff" the two apps? Something like:

* Unzip two `.apk` files to two folders
* Compare the two folders in your favorite diff tool, like [WinMerge](https://winmerge.org/)
* What is different?

I imagine what you are describing is possible but would require lots of MSBuild knowledge to make it work. Depending on what the diff is; I'm not sure I would recommend it.

Would you be better off fanning out your build into multiple lanes in parallel? Azure DevOps can do this pretty easily, for example ours looks like this:

![image](https://github.com/user-attachments/assets/18472271-e0cb-4e74-9471-824ef4efb2b4)

**Reply by FlavioGoncalves-Cayas** [reply]
I just took a look with FileMerge on MacOS. There are diffs in

- all .dex files
- all .so files
- assemblies/assemblies.blob
- some resources, even some which seem to be android system related, like drawable/$avd_hide_password__0.xml (maybe because of different colors in colors.xml?)

I guess it's not as easy as one might imagine.

> Would you be better off fanning out your build into multiple lanes in parallel? Azure DevOps can do this pretty easily, for example ours looks like this:

I wasn't necessarily looking to make it faster, but save costs and resources by removing redundant steps in the build.

**Reply by jonathanpeppers** [reply]
So many files are different, I'm not sure I would attempt it. The time you would spend to get it to work (in your salary, for example), seems like it would cost more than the saved CI $$ dollars.

You might consider just running the pipeline ~daily? Or on-demand? Easy way to save?
